upup/pkg/fi/secrets/clientset_secretstore.go

```diff
@@ -0,0 +1,185 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package secrets
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kops/pkg/apis/kops"
+	kopsinternalversion "k8s.io/kops/pkg/client/clientset_generated/clientset/typed/kops/internalversion"
+	"k8s.io/kops/pkg/pki"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/util/pkg/vfs"
+)
+
+// NamePrefix is a prefix we use to avoid collisions with other keysets
+const NamePrefix = "token-"
+
+type ClientsetSecretStore struct {
+	namespace string
+	clientset kopsinternalversion.KopsInterface
+}
+
+var _ fi.SecretStore = &ClientsetSecretStore{}
+
+// NewClientsetSecretStore is the constructor for ClientsetSecretStore
+func NewClientsetSecretStore(clientset kopsinternalversion.KopsInterface, namespace string) fi.SecretStore {
+	c := &ClientsetSecretStore{
+		clientset: clientset,
+		namespace: namespace,
+	}
+	return c
+}
+
+// FindSecret implements fi.SecretStore::FindSecret
+func (c *ClientsetSecretStore) FindSecret(name string) (*fi.Secret, error) {
+	s, err := c.loadSecret(name)
+	if err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+// ListSecrets implements fi.SecretStore::ListSecrets
+func (c *ClientsetSecretStore) ListSecrets() ([]string, error) {
+	list, err := c.clientset.Keysets(c.namespace).List(v1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error listing keysets: %v", err)
+	}
+
+	var names []string
+	for i := range list.Items {
+		keyset := &list.Items[i]
+
+		switch keyset.Spec.Type {
+		case kops.SecretTypeSecret:
+			name := strings.TrimPrefix(keyset.Name, NamePrefix)
+			names = append(names, name)
+		}
+	}
+
+	return names, nil
+}
+
+// Secret implements fi.SecretStore::Secret
+func (c *ClientsetSecretStore) Secret(name string) (*fi.Secret, error) {
+	s, err := c.FindSecret(name)
+	if err != nil {
+		return nil, err
+	}
+	if s == nil {
+		return nil, fmt.Errorf("Secret not found: %q", name)
+	}
+	return s, nil
+}
+
+// DeleteSecret implements fi.SecretStore::DeleteSecret
+func (c *ClientsetSecretStore) DeleteSecret(item *fi.KeystoreItem) error {
+	client := c.clientset.Keysets(c.namespace)
+	return fi.DeleteKeysetItem(client, item.Name, kops.SecretTypeKeypair, item.Id)
+}
+
+// GetOrCreateSecret implements fi.SecretStore::GetOrCreateSecret
+func (c *ClientsetSecretStore) GetOrCreateSecret(name string, secret *fi.Secret) (*fi.Secret, bool, error) {
+	for i := 0; i < 2; i++ {
+		s, err := c.FindSecret(name)
+		if err != nil {
+			return nil, false, err
+		}
+
+		if s != nil {
+			return s, false, nil
+		}
+
+		_, err = c.createSecret(secret, name)
+		if err != nil {
+			if errors.IsAlreadyExists(err) && i == 0 {
+				glog.Infof("Got already-exists error when writing secret; likely due to concurrent creation.  Will retry")
```

Comment:
Anyway to test for this, than assume that this is the problem?

Reply:
We just read it, it didn't exist; we try to create it, it does exist.  Presumably somebody created it in between the read & write.  Or we're using an inconsistent datastore ;-)  The only thing we can do is to retry...